packages/datadog-instrumentations/src/ioredis.js

```diff
@@ -0,0 +1,85 @@
+'use strict'
+
+const {
+  channel,
+  addHook
+} = require('./helpers/instrument')
+const shimmer = require('../../datadog-shimmer')
+
+const startCh = channel('apm:ioredis:command:start')
+const asyncEndCh = channel('apm:ioredis:command:async-end')
+const endCh = channel('apm:ioredis:command:end')
+const errorCh = channel('apm:ioredis:command:error')
+
+addHook({ name: 'ioredis', versions: ['>=2'] }, Redis => {
+  shimmer.wrap(Redis.prototype, 'sendCommand', sendCommand => function (command, stream) {
+    if (!startCh.hasSubscribers) return sendCommand.apply(this, arguments)
+
+    if (!command || !command.promise) return sendCommand.apply(this, arguments)
+
+    const options = this.options || {}
+    const connectionName = this.options.connectionName
+    const db = options.db
+    const connectionOptions = { host: options.host, port: options.port }
+    startCh.publish({ db, command: command.name, args: command.args, connectionOptions, connectionName })
+
+    try {
+      wrap(asyncEndCh, errorCh, command.promise)
+
+      return sendCommand.apply(this, arguments)
+    } catch (err) {
+      err.stack // trigger getting the stack at the original throwing point
+      errorCh.publish(err)
+
+      throw err
+    } finally {
+      endCh.publish(undefined)
+    }
+  })
+  return Redis
+})
+
+function wrap (asyncEndCh, errorCh, done) {
+  if (typeof done === 'function' || !done) {
```

Comment:
The type should be known in advance, so should this be split in separate methods like `wrapCallback` and `wrapPromise`? Also, I think it's used only once and for a promise, so are wrap functions even needed at all?